client/config.go

```diff
@@ -0,0 +1,101 @@
+package client
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/cloudquery/cloudquery/plugin"
+	"github.com/rs/zerolog/log"
+	"gopkg.in/yaml.v3"
+	"io/ioutil"
+	"os"
+)
+
+// Configuration provider header
+const headerConfig = `providers:`
+
+type Config struct {
+	Providers []ProviderData
+}
+
+type ProviderData struct {
+	Name    string
+	Version string `default:"latest"`
+	Rest    map[string]interface{} `yaml:",inline"`
+}
+
+func Init(configPath string) error {
+
+	log.Debug().Str("path", configPath).Msg("reading configuration file")
+	data, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return err
+	}
+
+	config := Config{}
+	err = yaml.Unmarshal(data, &config)
+	if err != nil {
+		return err
+	}
+
+	workingDir, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	cqProvidersDir := workingDir + "/.cq/providers/"
+	err = os.MkdirAll(cqProvidersDir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	for _, provider := range config.Providers {
+		if provider.Name == "" {
+			return fmt.Errorf("provider must contain key: name")
+		}
+		log.Debug().Str("path", configPath).Msg("reading configuration file")
+		err := downloadProviderPlugin(provider.Name, provider.Version)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func GenerateConfig(configPath string, providers []string, force bool) error {
+	var s bytes.Buffer
+	_, err := s.WriteString(headerConfig)
+	if err != nil {
+		return err
+	}
+
+	for _, provider := range providers {
+		err := downloadProviderPlugin(provider, "latest")
+		if err != nil {
+			return err
+		}
+		p, err := plugin.GetManager().GetOrCreateProvider(provider, "latest")
+		if err != nil {
+			return err
+		}
+		log.Debug().Str("provider", provider).Msg("Building provider configuration yaml")
+		configYaml, err := p.GenConfig()
+		if err != nil {
+			_ = plugin.GetManager().KillProvider(provider)
```

Comment:
I agree that we need to return the first error but I'd at least print this error as well.

Reply:
refactoring this in a different branch